### ISSUE TITLE
Cycles between Dart and Java memory

### ISSUE THREAD

**Issue by dcharkes** [dart-lang/jnigen#370]
Because the Dart and Java GCs do not know about each other, we can end up in situations where we lose reference in the Dart to a Dart object A and lose reference in Java to a Java object B, but B holds on to A via a (for example via an open port if A is a closure) and A holds on to B via a JObject.

It is not clear if this issue is going to be very common, but we should keep an eye on it: We should document any patterns that we find that might result in this. Probably we can construct one with https://github.com/dart-lang/jnigen/pull/369.

If it's common we should probably provide some kind of weak reference system in at least one direction. It might depend per use case if this is possible.

(A completely out of scope, but conceptually cool, solution: A unified Dart and Java heap in the style of Oilpan which unifies the C++ DOM heap with the JS heap: https://v8.dev/blog/oilpan-library)

**Comment by HosseinYousefi** [comment]
I don't think this is possible actually. Everything happens based on Java global references. If a JObject is unreachable in Dart, the underlying global ref will be deleted and if the same object had been unreachable in Java as well, it will be GCd.

Can you construct an example where this can happen?

**Comment by dcharkes** [comment]
How about of we create an interface implementation, that in one of it's closures holds on to a JObject that has a field. And then we store our just created implementation into that field?

**Comment by HosseinYousefi** [comment]
Right!

```dart
Foo foo;
foo.bar = Interface.implement($InterfaceImpl(
  f: () {
    return foo;
  }
));
```

Now when `foo` is lost, it's still holding `.bar` which holds `foo`.
